### ISSUE TITLE
Callback protocol with `__call__` overloads raises `reportGeneralTypeIssues`

### ISSUE THREAD

**Issue by simensol** [microsoft/pyright#4181]
**Describe the bug**
Callback protocol with `__call__` overloads raises `reportGeneralTypeIssues`.

**To Reproduce**
I have implemented a [Callback protocol](https://peps.python.org/pep-0544/#callback-protocols) with two `@overload`:

```python
from typing import Protocol, overload


class FooBarCallable(Protocol):
    @overload
    def __call__(self, x: str) -> str:
        ...

    @overload
    def __call__(self, x: int) -> int:
        ...

    def __call__(self, x: str | int) -> str | int:
        ...


def foo(x: str):
    return x


def bar(x: int):
    return x


def baz(x: float):
    return x


# Type check foo():
_: FooBarCallable = foo


# Type check bar():
_: FooBarCallable = bar


# Type check baz():
_: FooBarCallable = baz

```

**Expected behavior**
I expect type check of `foo()` and `bar()` to pass, but `baz()` to fail.

**Actual behavior**
Type check of all three methods `foo()`, `bar()` and `baz()` fail:

```bash
pyright 1.1.279
test.py
  test.py:30:21 - error: Expression of type "(x: str) -> str" cannot be assigned to declared type "FooBarCallable"
    One or more overloads of "__call__" is not assignable
      Type "(x: str) -> str" cannot be assigned to type "(x: int) -> int"
        Parameter 1: type "int" cannot be assigned to type "str"
          "int" is incompatible with "str"
        Function return type "str" is incompatible with type "int"
          "str" is incompatible with "int" (reportGeneralTypeIssues)
  test.py:34:21 - error: Expression of type "(x: int) -> int" cannot be assigned to declared type "FooBarCallable"
    One or more overloads of "__call__" is not assignable
      Type "(x: int) -> int" cannot be assigned to type "(x: str) -> str"
        Parameter 1: type "str" cannot be assigned to type "int"
          "str" is incompatible with "int"
        Function return type "int" is incompatible with type "str"
          "int" is incompatible with "str" (reportGeneralTypeIssues)
  test.py:38:21 - error: Expression of type "(x: float) -> float" cannot be assigned to declared type "FooBarCallable"
    One or more overloads of "__call__" is not assignable
      Type "(x: float) -> float" cannot be assigned to type "(x: str) -> str"
        Parameter 1: type "str" cannot be assigned to type "float"
          "str" is incompatible with "float"
        Function return type "float" is incompatible with type "str"
          "float" is incompatible with "str" (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations 
Completed in 1.668sec
```

**Comment by erictraut** [comment]
This error is correct. The protocol indicates that the `__call__` should accept either a `str` or an `int` input, but the functions `foo`, `bar`, and `baz` do not satisfy this requirement. They are therefore incompatible with the protocol.

If your intent is to support a call that can accept either a `str` or an `int`, then you would need to create two different protocols and create a union of the two.
